### ISSUE TITLE
[ocaml5-issue] Segfault in Buffer test (5.0, 32bit)

### ISSUE THREAD

**Issue by shym** [ocaml-multicore/multicoretests#306]
Observed in a run on released OCaml 5.0 in 32 bits, a SEGV in the buffer test:
https://github.com/shym/multicoretests/actions/runs/4367430742/jobs/7638729780#step:20:1204

```
random seed: 470817338
generated error fail pass / total     time test name

[ ]    0    0    0    0 / 1000     0.0s STM Buffer test sequential
[ ]    0    0    0    0 / 1000     0.0s STM Buffer test sequential (generating)
[✓] 1000    0    0 1000 / 1000     1.9s STM Buffer test sequential

File "src/buffer/dune", line 4, characters 7-16:
4 |  (name stm_tests)
           ^^^^^^^^^
[ ]    0    0    0    0 / 1000     0.0s STM Buffer test parallel
(cd _build/default/src/buffer && ./stm_tests.exe --verbose)
Command got signal SEGV.
```

**Comment by jmid** [comment]
Cool! :sunglasses: :+1: 
I suggest hacking the tested property to print the triple as a side-effect.
That way, you can get a better sense of the input causing it - and perhaps write an STM-independent reproducable.

The old `diff` from my last hack follows below.
(we should consider incorporating some way to obtain something like it that doesn't involve hacking on the library...)
```
diff --git a/lib/STM_domain.ml b/lib/STM_domain.ml
index b8bd1c7..fc11554 100644
--- a/lib/STM_domain.ml
+++ b/lib/STM_domain.ml
@@ -17,8 +17,11 @@ module Make (Spec: Spec) = struct
     let res_arr = Array.map (fun c -> Domain.cpu_relax(); Spec.run c sut) cs_arr in
     List.combine cs (Array.to_list res_arr)
 
-  let agree_prop_par (seq_pref,cmds1,cmds2) =
+  let ch = open_out "input_log.txt"
+
+  let agree_prop_par ((seq_pref,cmds1,cmds2) as trip) =
     assume (all_interleavings_ok seq_pref cmds1 cmds2 Spec.init_state);
+    Printf.fprintf ch "Triple: %s\n%!" Print.(triple (list Spec.show_cmd) (list Spec.show_cmd) (list Spec.show_cmd) trip);
     let sut = Spec.init_sut () in
     let pref_obs = interp_sut_res sut seq_pref in
     let wait = Atomic.make true in
@@ -29,7 +32,11 @@ module Make (Spec: Spec) = struct
     let obs1 = match obs1 with Ok v -> v | Error exn -> raise exn in
     let obs2 = match obs2 with Ok v -> v | Error exn -> raise exn in
     let ()   = Spec.cleanup sut in
-    check_obs pref_obs obs1 obs2 Spec.init_state
+    let res = check_obs pref_obs obs1 obs2 Spec.init_state in
+    if res
+    then Printf.fprintf ch "- true\n%!"
+    else Printf.fprintf ch "- false\n%!";
+    res
       || Test.fail_reportf "  Results incompatible with linearized model\n\n%s"
          @@ print_triple_vertical ~fig_indent:5 ~res_width:35
            (fun (c,r) -> Printf.sprintf "%s : %s" (Spec.show_cmd c) (show_res r))
```

**Comment by dra27** [comment]
Is the crash likely to be more to do with 32-bit or more to do with bytecode? (I can't remember if you're testing both 32 and 64-bit bytecode at the moment?)

**Comment by jmid** [comment]
> Is the crash likely to be more to do with 32-bit or more to do with bytecode? (I can't remember if you're testing both 32 and 64-bit bytecode at the moment?)

We have separate `bytecode` (64-bit) and `32-bit` targets and haven't observed this before under bytecode AFAIR.

**Comment by shym** [comment]
I could find a seed that will give the assertion error also in 64-bit bytecode, but I don’t think I’ve seen the SEGV (which means nothing, of course).
```
$ ocamlrund _build/default/src/buffer/stm_tests.bc -v -s 99856410
### OCaml runtime: debug mode ###
random seed: 99856410
generated error fail pass / total     time test name
[ ]    0    0    0    0 / 1000     2.9s STM Buffer test parallel (shrinking:   13.0005)[00] file runtime/str.c; line 44 ### Assertion failed: Byte (s, temp - Byte (s, temp)) == 0
```

**Comment by shym** [comment]
Probably fixed by ocaml/ocaml#12104, as I could not reproduce it with that patch in 50 runs with the seed made it fail reliably often and 400 other random runs.

**Comment by jmid** [comment]
Thanks! That one fortunately turned out to be a quick and easy fix :+1: 
We should remember to add it to the list of issues in the README...
